multipath.go

```diff
@@ -0,0 +1,136 @@
+// frame is not acked in time, it is sent over another subflow, until all
+// available subflows have been tried. Payload size and frame number uses
+// variable-length integer encoding as described here: https:
+// //tools.ietf.org/html/draft-ietf-quic-transport-29#section-16
+//
+//       --------------------------------------------------------
+//      |  payload size(1-8)  |  frame number (1-8)  |  payload  |
+//       --------------------------------------------------------
+//
+//       ---------------------------------------
+//      |  00000000  |  ack frame number (1-8)  |
+//       ---------------------------------------
+//
+// Ack frames with frame number < 10 are reserved for control. For now only 0
+// and 1 are used, for ping and pong frame respectively. They are for updating
+// RTT on inactive subflows and detecting recovered subflows.
+//
+// Ping frame:
+//       -------------------------
+//      |  00000000  |  00000000  |
+//       -------------------------
+//
+// Pong frame:
+//       -------------------------
+//      |  00000000  |  00000001  |
+//       -------------------------
+//
+
+package multipath
+
+import (
+	"bytes"
+	"errors"
+	"sync/atomic"
+	"time"
+
+	"github.com/getlantern/golog"
+	"github.com/google/uuid"
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+const (
+	minFrameNumber uint64 = 10
+	frameTypePing  uint64 = 0
+	frameTypePong  uint64 = 1
+
+	maxFrameSizeToCalculateRTT int = 1500
+	leadBytesLength                = 1 + 16 // 1 byte version + 16 bytes CID
+	recieveQueueLength             = 4096
+	maxVarIntLength                = 8
+	probeInterval                  = time.Minute
+	longRTT                        = time.Minute
+	rttAlpha                       = 0.5 // this causes EMA to reflect changes more rapidly
+)
+
+var (
+	ErrUnexpectedVersion = errors.New("unexpected version")
+	ErrUnexpectedCID     = errors.New("unexpected connnection ID")
+	ErrClosed            = errors.New("closed connection")
+	ErrFailOnAllDialers  = errors.New("fail on all dialers")
+	log                  = golog.LoggerFor("multipath")
+	zeroCID              connectionID
+)
+
+type connectionID uuid.UUID
+type frame struct {
+	fn    uint64
+	bytes []byte
+}
+
+type sendFrame struct {
+	fn              uint64
+	sz              uint64
+	buf             []byte
+	released        *int32 // 1 == true; 0 == false. Use pointer so copied object still references the same address, as buf does
+	retransmissions int
+}
+
+func composeFrame(fn uint64, b []byte) *sendFrame {
+	sz := len(b)
+	buf := pool.Get(maxVarIntLength + maxVarIntLength + sz)
+	wb := bytes.NewBuffer(buf[:0])
+	WriteVarInt(wb, uint64(sz))
+	WriteVarInt(wb, fn)
+	if sz > 0 {
+		wb.Write(b)
+	}
+	var released int32
+	return &sendFrame{fn: fn, sz: uint64(sz), buf: wb.Bytes(), released: &released}
+}
+
+func (f *sendFrame) isDataFrame() bool {
+	return f.sz > 0
+}
+
+func (f *sendFrame) release() {
+	if atomic.CompareAndSwapInt32(f.released, 0, 1) {
```

Comment:
Initially there's no such protection to release buffer more than once here, and later on as part of the failed attempt to locate the data corruption I added this, but the error never showed up. I also doubled checked every place calling it. I tend to remove it altogether consider its high frequency. How do you think?

BTW thanks for benchmarking! Never thought that the performance is so close.

Reply:
Hmm I would be inclined to keep it.  It would help protect against bugs introduced by future work in this project and the few nanoseconds it might add doesn't seem like much of a price to pay.

Reply:
Makes sense. I think I'll keep it as is because with `sync.Once` we can not have error log if it's called more than once, which means a program error.